fvm/handler/contract.go

```diff
@@ -0,0 +1,153 @@
+package handler
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/onflow/cadence/runtime"
+
+	"github.com/onflow/flow-go/fvm/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type ContractHandler struct {
+	accounts                    *state.Accounts
+	draftUpdates                map[string]contractUpdate
+	restrictedDeploymentEnabled bool
+	authorizedAccounts          []runtime.Address
+	// handler doesn't have to be thread safe and right now
+	// is only used in a single thread but a mutex has been added
+	// here to prevent accidental multi-thread use in the future
+	lock sync.Mutex
+}
+
+func NewContractHandler(accounts *state.Accounts, restrictedDeploymentEnabled bool, authorizedAccounts []runtime.Address) *ContractHandler {
+	return &ContractHandler{
+		accounts:                    accounts,
+		draftUpdates:                make(map[string]contractUpdate),
+		restrictedDeploymentEnabled: restrictedDeploymentEnabled,
+		authorizedAccounts:          authorizedAccounts,
+	}
+}
+
+func (h *ContractHandler) GetContractNames(address runtime.Address) (names []string, err error) {
+	names, err = h.accounts.GetContractNames(flow.Address(address))
+	return
+}
+
+func (h *ContractHandler) GetContract(address runtime.Address, name string) (code []byte, err error) {
+	code, err = h.accounts.GetContract(name, flow.Address(address))
+	return
+}
+
+func (h *ContractHandler) SetContract(address runtime.Address, name string, code []byte, signingAccounts []runtime.Address) (err error) {
+	// check if authorized
+	if !h.isAuthorized(signingAccounts) {
+		return errors.New("code deployment requires authorization from specific accounts")
+	}
+	add := flow.Address(address)
+	// check if account exists
+	ok, err := h.accounts.Exists(add)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("account with address %s does not exist", address)
+	}
```

Comment:
I think this defensive check is not required, it's probably safe to remove it

Reply:
removed them